samwise.js

```diff
@@ -5,53 +5,56 @@
  * Released under the MIT license.
```

Comment:
SCRIPT-This function seems to have grown quite complex. Can we break it down further?